*** includes/specials/SpecialUpload.php	2008-12-23 17:19:14.000000000 -0600
--- /var/www/html/mediawiki-1.14.0/includes/specials/SpecialUpload.php	2009-04-14 14:58:20.000000000 -0500
***************
*** 272,277 ****
--- 272,282 ----
  		$this->cleanupTempFile();
  	}
  
+ 	function showSuccess() {
+ 		global $wgOut;
+ 		$wgOut->redirect( $this->mLocalFile->getTitle()->getFullURL() );
+ 	}
+ 
  	/**
  	 * Do the upload
  	 * Checks are made in SpecialUpload::execute()
***************
*** 286,292 ****
  
  	 	switch($value) {
  			case self::SUCCESS:
! 				$wgOut->redirect( $this->mLocalFile->getTitle()->getFullURL() );
  				break;
  
  			case self::BEFORE_PROCESSING:
--- 291,297 ----
  
  	 	switch($value) {
  			case self::SUCCESS:
! 				$this->showSuccess();
  				break;
  
  			case self::BEFORE_PROCESSING:
