Index: GlobalFunctions.php
===================================================================
--- GlobalFunctions.php	(revision 53111)
+++ GlobalFunctions.php	(working copy)
@@ -2934,6 +2934,6 @@
  */
 function wfStripIllegalFilenameChars( $name ) {
 	$name = wfBaseName( $name );
-	$name = preg_replace ( "/[^".Title::legalChars()."]|:/", '-', $name );
+	$name = preg_replace ( "/[^".Title::legalChars()."]/", '-', $name );
 	return $name;
 }
