Index: includes/sparql/SMW_SparqlDatabase4Store.php
===================================================================
--- includes/sparql/SMW_SparqlDatabase4Store.php	(revision 38933)
+++ includes/sparql/SMW_SparqlDatabase4Store.php	(working copy)
@@ -1,19 +1,19 @@
 <?php
 /**
  * 4Store specific adjustments for SMWSparqlDatabase
- * 
+ *
  * @file
  * @ingroup SMWSparql
- * 
+ *
  * @author Markus Krötzsch
  */
 
 /**
  * Specific modifications of the SPARQL database implementation for 4Store.
- * 
+ *
  * @since 1.6
  * @ingroup SMWSparql
- * 
+ *
  * @author Markus Krötzsch
  */
 class SMWSparqlDatabase4Store extends SMWSparqlDatabase {
@@ -39,6 +39,7 @@
 		curl_setopt( $this->m_curlhandle, CURLOPT_POST, true );
 		$parameterString = "query=" . urlencode( $sparql ) . "&restricted=1";
 		curl_setopt( $this->m_curlhandle, CURLOPT_POSTFIELDS, $parameterString );
+		curl_setopt( $this->m_curlhandle, CURLOPT_PROXY, false );
 
 		$xmlResult = curl_exec( $this->m_curlhandle );
 
@@ -88,7 +89,7 @@
 	 * SMWSparqlDatabase::throwSparqlErrors(). If errors occur and this
 	 * method does not throw anything, then an empty result with an error
 	 * code is returned.
-	 * 
+	 *
 	 * This method is specific to 4Store since it uses POST parameters that
 	 * are not given in the specification.
 	 *
@@ -114,4 +115,4 @@
 		}
 	}
 
-}
\ No newline at end of file
+}
Index: SMW_Settings.php
===================================================================
--- SMW_Settings.php	(revision 38934)
+++ SMW_Settings.php	(working copy)
@@ -23,7 +23,7 @@
 # web. Change it if required ($wgScriptPath is the path to the base directory
 # of your wiki). No final slash.
 ##
-$smwgScriptPath = ( 
+$smwgScriptPath = (
 	( version_compare( $wgVersion, '1.16', '>=' ) && isset( $wgExtensionAssetsPath ) && $wgExtensionAssetsPath )
 	? $wgExtensionAssetsPath : $wgScriptPath . '/extensions'
 	) . '/SemanticMediaWiki';
@@ -65,6 +65,9 @@
 $smwgSparqlDataEndpoint = 'http://localhost:8080/data/';
 ##
 
+// Wikia overrides
+require_once( 'Wikia_Settings.php' );
+
 // load global constants and setup functions
 require_once( 'includes/SMW_Setup.php' );
 
