Index: maintenance/updaters.inc
===================================================================
--- maintenance/updaters.inc	(revision 53499)
+++ maintenance/updaters.inc	(working copy)
@@ -884,7 +884,7 @@
  * @bug 3946
  */
 function do_page_random_update() {
-	global $wgDatabase;
+	/*global $wgDatabase;
 
 	wfOut( "Setting page_random to a random value on rows where it equals 0..." );
 
@@ -892,7 +892,7 @@
 	$wgDatabase->query( "UPDATE $page SET page_random = RAND() WHERE page_random = 0", 'do_page_random_update' );
 	$rows = $wgDatabase->affectedRows();
 
-	wfOut( "changed $rows rows\n" );
+	wfOut( "changed $rows rows\n" );*/
 }
 
 function do_templatelinks_update() {
@@ -1211,7 +1211,7 @@
 }
 
 function do_category_population() {
-	if( update_row_exists( 'populate category' ) ) {
+	/*if( update_row_exists( 'populate category' ) ) {
 		wfOut( "...category table already populated.\n" );
 		return;
 	}
@@ -1221,7 +1221,7 @@
 "may want to hit Ctrl-C and do this manually with maintenance/\n".
 "populateCategory.php.\n";
 	populateCategory( '', 10, 0, true );
-	wfOut( "Done populating category table.\n" );
+	wfOut( "Done populating category table.\n" );*/
 }
 
 function do_populate_parent_id() {
