Index: includes/Parser.php
===================================================================
--- includes/Parser.php	(revision 14250)
+++ includes/Parser.php	(working copy)
@@ -3706,6 +3706,20 @@
 		$this->mFunctionHooks[$name] = $callback;
 		return $oldVal;
 	}
+	
+	
+	/**
+	 * Add an extension handler object which will be serialized with the
+	 * parser cache output. This can be used to add style sheets or links,
+	 * invalidate cached pages, etc.
+	 *
+	 * Expected to be used from extension hooks primarily.
+	 *
+	 * @fixme Document this interface
+	 */
+	public function addOutputHandler( $object ) {
+		$this->mOutput->addOutputHandler( $object );
+	}
 
 	/**
 	 * Replace <!--LINK--> link placeholders with actual links, in the buffer
@@ -4108,9 +4122,10 @@
 		$mTemplates,        # 2-D map of NS/DBK to ID for the template references. ID=zero for broken.
 		$mImages,           # DB keys of the images used, in the array key only
 		$mExternalLinks,    # External link URLs, in the key only
-		$mHTMLtitle,		# Display HTML title
-		$mSubtitle,			# Additional subtitle
-		$mNewSection;		# Show a new section link?
+		$mHTMLtitle,        # Display HTML title
+		$mSubtitle,         # Additional subtitle
+		$mNewSection,       # Show a new section link?
+		$mHandlers;         # Extension objects to run on output
 
 	function ParserOutput( $text = '', $languageLinks = array(), $categoryLinks = array(),
 		$containsOldMagic = false, $titletext = '' )
@@ -4129,6 +4144,7 @@
 		$this->mHTMLtitle = "" ;
 		$this->mSubtitle = "" ;
 		$this->mNewSection = false;
+		$this->mHandlers = array();
 	}
 
 	function getText()                   { return $this->mText; }
@@ -4155,6 +4171,8 @@
 	function addLanguageLink( $t )       { $this->mLanguageLinks[] = $t; }
 	function addExternalLink( $url )     { $this->mExternalLinks[$url] = 1; }
 	
+	function addOutputHandler( $object ) { $this->mHandlers[] = $object; }
+	
 	function setNewSection( $value ) {
 		$this->mNewSection = (bool)$value;
 	}
@@ -4197,6 +4215,18 @@
 		       !isset( $this->mVersion ) ||
 		       version_compare( $this->mVersion, MW_PARSER_VERSION, "lt" );
 	}
+	
+	/**
+	 * Apply any attached handler objects to the page output.
+	 * @param OutputPage $outputPage
+	 */
+	public function applyHandlers( $outputPage ) {
+		if( is_array( $this->mHandlers ) ) {
+			foreach( $this->mHandlers as $handler ) {
+				$handler->apply( $outputPage );
+			}
+		}
+	}
 }
 
 /**
Index: includes/OutputPage.php
===================================================================
--- includes/OutputPage.php	(revision 14250)
+++ includes/OutputPage.php	(working copy)
@@ -305,6 +305,7 @@
 			$this->mPagetitle = $parserOutput->mHTMLtitle ;
 			$this->mSubtitle .= $parserOutput->mSubtitle ;
 		}
+		$parserOutput->applyHandlers( $this );
 	}
 
 	function addParserOutput( &$parserOutput ) {
@@ -383,6 +384,7 @@
 			$this->addCategoryLinks( $parserOutput->getCategories() );
 			$this->addKeywords( $parserOutput );
 			$this->mNewSectionLink = $parserOutput->getNewSection();
+			$parserOutput->applyHandlers( $this );
 			$text = $parserOutput->getText();
 			wfRunHooks( 'OutputPageBeforeHTML', array( &$this, &$text ) );
 			$this->addHTML( $text );
